### DISCUSSION TITLE
How do you add variants to a new listing?

### DISCUSSION THREAD

**Discussion by geolectric** [etsy/open-api#766] [Q&A] [answered]
I'm trying to create many listings that will have the exact same variant choices.

So, I create a new listing:
```
data = {
    "quantity": quantity,
    "title": title,
    "description": description,
    "price": price,
    "who_made": who_made,
    "when_made": when_made,
    "taxonomy_id": taxonomy_id
    .....
}

resp = requests.post(
    url=f"https://openapi.etsy.com/v3/application/shops/{shop_id}/listings",
    headers=headers,
    data=data
)

resp_json = resp.json()
listing_id = resp_json["listing_id"]
```

So, now I have the listing_id, but how do I add the variants? It'd be convenient if we could just add them when creating the listing, but that doesn't seem to be an option?

Do I need to use updateListingInventory? What does updateListing vs updateListingProperty do?

Honestly, the API is a bit confusing on how everything ties together. The names are pretty generic and it's hard to know what's used for what...

What is a Product vs Offering?

I manually created a listing with the variants I want, and then ran getListingInventory on it, and this was the output:

```
{
    'products': [{
        'product_id': 12403675300,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12181384285,
            'quantity': 1,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 1000,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [103715606336],
            'values': ['Digital File']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1105921146363],
            'values': ['(Digital File)']
        }]
    }, {
        'product_id': 12066054717,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12181368411,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [103715606336],
            'values': ['Digital File']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683482737],
            'values': ['8 x 8']
        }]
    }, {
        'product_id': 12066054719,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12181368415,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 200,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [103715606336],
            'values': ['Digital File']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1069468004266],
            'values': ['10 x 10']
        }]
    }, {
        'product_id': 12403659412,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12367411666,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 300,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [103715606336],
            'values': ['Digital File']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683926753],
            'values': ['12 x 12']
        }]
    }, {
        'product_id': 12403675304,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12181384291,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [294355636092],
            'values': ['Basic Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1105921146363],
            'values': ['(Digital File)']
        }]
    }, {
        'product_id': 12395018480,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918128,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [294355636092],
            'values': ['Basic Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683482737],
            'values': ['8 x 8']
        }]
    }, {
        'product_id': 12057358761,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12172525065,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 200,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [294355636092],
            'values': ['Basic Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1069468004266],
            'values': ['10 x 10']
        }]
    }, {
        'product_id': 12395018486,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918132,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 300,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [294355636092],
            'values': ['Basic Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683926753],
            'values': ['12 x 12']
        }]
    }, {
        'product_id': 12403675312,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12367427338,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [130071027068],
            'values': ['Fine Art Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1105921146363],
            'values': ['(Digital File)']
        }]
    }, {
        'product_id': 12057358769,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918136,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 400,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [130071027068],
            'values': ['Fine Art Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683482737],
            'values': ['8 x 8']
        }]
    }, {
        'product_id': 12395018492,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918140,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 500,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [130071027068],
            'values': ['Fine Art Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1069468004266],
            'values': ['10 x 10']
        }]
    }, {
        'product_id': 12395018494,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918142,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 600,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [130071027068],
            'values': ['Fine Art Print']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683926753],
            'values': ['12 x 12']
        }]
    }, {
        'product_id': 12403675324,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12367427352,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [111200009186],
            'values': ['Framed Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1105921146363],
            'values': ['(Digital File)']
        }]
    }, {
        'product_id': 12395018496,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918146,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 700,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [111200009186],
            'values': ['Framed Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683482737],
            'values': ['8 x 8']
        }]
    }, {
        'product_id': 12395018498,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12172525077,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 800,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [111200009186],
            'values': ['Framed Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1069468004266],
            'values': ['10 x 10']
        }]
    }, {
        'product_id': 12395018500,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12172525079,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 900,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [111200009186],
            'values': ['Framed Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683926753],
            'values': ['12 x 12']
        }]
    }, {
        'product_id': 12403675340,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12181384309,
            'quantity': 1,
            'is_enabled': False,
            'is_deleted': False,
            'price': {
                'amount': 100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [113096407979],
            'values': ['Stretched Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1105921146363],
            'values': ['(Digital File)']
        }]
    }, {
        'product_id': 12395018502,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12172525083,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 1000,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [113096407979],
            'values': ['Stretched Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683482737],
            'values': ['8 x 8']
        }]
    }, {
        'product_id': 12057358781,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918152,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 1100,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [113096407979],
            'values': ['Stretched Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1069468004266],
            'values': ['10 x 10']
        }]
    }, {
        'product_id': 12057358785,
        'sku': '',
        'is_deleted': False,
        'offerings': [{
            'offering_id': 12358918156,
            'quantity': 5,
            'is_enabled': True,
            'is_deleted': False,
            'price': {
                'amount': 1200,
                'divisor': 100,
                'currency_code': 'USD'
            }
        }],
        'property_values': [{
            'property_id': 510,
            'property_name': 'Style',
            'scale_id': None,
            'scale_name': None,
            'value_ids': [113096407979],
            'values': ['Stretched Canvas']
        }, {
            'property_id': 501,
            'property_name': 'Dimensions',
            'scale_id': 344,
            'scale_name': ' inches',
            'value_ids': [1092683926753],
            'values': ['12 x 12']
        }]
    }],
    'price_on_property': [510, 501],
    'quantity_on_property': [510, 501],
    'sku_on_property': [],
    'listing': None
}
```

So, basically, I just want to add that exact same thing to all the listings I create...

What's the workflow for this?

Thanks

**Comment by etsyachristensen** [comment, marked as answer]
You have the basics so far. You do in fact need to run updateListingInventory. You are half way there in that you've got a good set of inventory data. Your next step will be to remove the information that does not get re-submitted to the listing and modify the price data to be a float and not an object. See the tutorial on handling updating inventory here: https://developers.etsy.com/documentation/tutorials/listings#updating-inventory

**Reply by geolectric** [reply]
@etsyachristensen thanks for your replies on here! I didn't even notice the tutorials, so those will be super helpful. I'll look through them all now. Thanks
